src/org/jetbrains/java/decompiler/modules/decompiler/TryHelper.java

```diff
@@ -185,4 +182,69 @@ private static boolean collapseTry(CatchStatement catchStat) {
     }
     return false;
   }
+
+  // try {
+  //   ...
+  // } catch (Exception e) {
+  //   ...
+  // } catch (Exception e) {
+  //   ...
+  // }
+  //
+  // into
+  // try {
+  //   try {
+  //     ...
+  //   } catch (Exception e) {
+  //     ...
+  //   }
+  // } catch (Exception e) {
+  //   ...
+  // }
+  public static boolean splitTryWithSameCatch(RootStatement root) {
+    return splitTryWithSameCatch(root, root);
+  }
+
+  private static boolean splitTryWithSameCatch(RootStatement root, Statement stat) {
+    boolean ret = false;
+
+    for (Statement st : new ArrayList<>(stat.getStats())) {
+      ret |= splitTryWithSameCatch(root, st);
+    }
+
+    if (stat.type == Statement.TYPE_TRYCATCH && ((CatchStatement)stat).getTryType() == CatchStatement.NORMAL) {
+      CatchStatement catchStat = (CatchStatement) stat;
+
+      Set<List<String>> setExec = new HashSet<>(catchStat.getExctStrings());
+
+      // Check if list of catch headers has a duplicate type
+      if (setExec.size() < catchStat.getExctStrings().size()) {
+        // To not have to calculate catch dominators, split along every catch type for now
+        // TODO: better impl that only splits the needed type
+
+        // Stack of nested statements
+        List<Statement> catchStack = new ArrayList<>();
```

Comment:
Looks like you only use the last element/top of the stack, so this could probably be replaced with just `Statement`